spec/requests/topics_controller_spec.rb

```diff
@@ -4,158 +4,1474 @@
+          private_topic: 403,
+          deleted_topic: 200,
+          deleted_secure_topic: 403,
+          deleted_private_topic: 403,
+          nonexist: 404
+        }
+        include_examples "various scenarios", expected
+      end
+
+      context 'admin' do
+        before do
+          sign_in(Fabricate(:admin))
+        end
+
+        expected = {
+          normal_topic: 200,
+          secure_topic: 200,
+          private_topic: 200,
+          deleted_topic: 200,
+          deleted_secure_topic: 200,
+          deleted_private_topic: 200,
+          nonexist: 404
+        }
+        include_examples "various scenarios", expected
+      end
+    end
+
+    it 'records a view' do
+      expect do
+        get "/t/#{topic.slug}/#{topic.id}.json"
+      end.to change(TopicViewItem, :count).by(1)
+    end
+
+    it 'records incoming links' do
+      user = Fabricate(:user)
+
+      expect do
+        get "/t/#{topic.slug}/#{topic.id}", params: {
+          u: user.username
+        }
+      end.to change { IncomingLink.count }.by(1)
+    end
+
+    context 'print' do
+      it "doesn't renders the print view when disabled" do
+        SiteSetting.max_prints_per_hour_per_user = 0
+
+        get "/t/#{topic.slug}/#{topic.id}/print"
+
+        expect(response).to be_forbidden
+      end
+
+      it 'renders the print view when enabled' do
+        SiteSetting.max_prints_per_hour_per_user = 10
+
+        get "/t/#{topic.slug}/#{topic.id}/print"
+
+        expect(response).to be_successful
+      end
+    end
+
+    it 'records redirects' do
+      get "/t/#{topic.id}", headers: { HTTP_REFERER: "http://twitter.com" }
+      get "/t/#{topic.slug}/#{topic.id}", headers: { HTTP_REFERER: nil }
+
+      link = IncomingLink.first
+      expect(link.referer).to eq('http://twitter.com')
+    end
+
+    it 'tracks a visit for all html requests' do
+      sign_in(user)
+      get "/t/#{topic.slug}/#{topic.id}"
+      topic_user = TopicUser.where(user: user, topic: topic).first
+      expect(topic_user.last_visited_at).to eq(topic_user.first_visited_at)
+    end
+
+    context 'consider for a promotion' do
+      before do
+        SiteSetting.tl1_requires_topics_entered = 0
+        SiteSetting.tl1_requires_read_posts = 0
+        SiteSetting.tl1_requires_time_spent_mins = 0
+        SiteSetting.tl1_requires_time_spent_mins = 0
+      end
+
+      it "reviews the user for a promotion if they're new" do
+        sign_in(user)
+        user.update_column(:trust_level, TrustLevel[0])
+        get "/t/#{topic.slug}/#{topic.id}.json"
+        user.reload
+        expect(user.trust_level).to eq(1)
+      end
+    end
+
+    context 'filters' do
+      def extract_post_stream
+        json = JSON.parse(response.body)
+        json["post_stream"]["posts"].map { |post| post["id"] }
+      end
+
+      before do
+        @post_ids = topic.posts.pluck(:id)
+        22.times do
+          @post_ids << Fabricate(:post, topic: topic).id
```

Comment:
Thanks for the feedback @tgxworld. I've updated the PR according to your feedback. Could you tell me if it's OK to do a loop like this here?

Reply:
Hmm can you try benchmarking how long it takes to run this spec? I'm wondering if we can change the chunk size somehow

Reply:
yeah forcing a chunk of 22 seems so overkill we should run it on a chunk of 5 ... also 1 spec is enough here, no need to do this very expensive creation destruction 5 times.